### PR TITLE
Plando support

### LINKED ISSUE #96: Make it possible to fully customize a seed
Unlike the current item placement options, this mode should override the filling algorithm completely — you're expected to manually customize a seed in its entirety.

This could be achieved e.g. by uploading an edited spoiler log? 

In addition to making "plando" seeds possible, this should also facilitate version-agnostic seed sharing, as the logic (which is version-dependent) isn't taken into account.

## Requirements

- [x] All items, including junk, should be placeable at any location.
  - This includes keys, maps and compasses (but not keycards unless keysanity is enabled).
- [x] Dungeon rewards (pendant/crystal) should be customizable.
- [x] Show a warning (either before or after generating the seed this way) that this seed overrides logic and may or may not be completable. 

### Nice to have
- [x] Export plando configuration from seed.
- [ ] Ability to customize enemy drops.
  - If going by the "upload a spoiler log" solution, we should write drop RNG to the spoiler log.

### PR DESCRIPTION
**Pull request by Vivelin**
I think I've got working plando support. It's still a little hacky, but the basis is here.

Basically, when generating a new seed, a plando config YAML file will be generated alongside the spoiler log. This file can be edited and used to start a new plando seed. For example, I generated a seed with only Big 20s.

I haven't done proper testing yet, and I still want to add support for changing enemy drops, but it should be fine in its current state. I might just postpone that as a nice-to-have.

Resolves #96

### CODE REVIEW

**Comment by MattEqualsCoder** [comment]
Did some testing, and unfortunately it seems like tracker doesn't quite work with plando. For me, it seemed like she thought things were at a completely different location. My guess is that the issue is that currently for regular seeds, things aren't saved into the database until you actually open tracker and save. When launching tracker, if there isn't data saved to the database already, the seed is basically regenerated.

My guess is that it's using the plando filename that seems to be used for the seed for that regeneration. Either we need to save the tracker session immediately upon plando generation, or we need to regenerate using the yaml file (though I'm a bit worried about that in the off change the file was deleted).

**Comment by Vivelin** [comment]
I guess we could just overhaul the database and actually store everything instead of just the seed. Like I think I left in a comment, maybe the PlandoConfig class could be repurposed for more?

We've overhauled so much at this point we're clashing a lot more with old vanilla SMZ3 stuff.

**Comment by MattEqualsCoder** [comment]
Yeah, I think currently it does save what's at locations IIRC, but I'm pretty sure it doesn't save things like what rewards are at dungeons or what medallions are for TR and MM. It just saves what the user thinks they are. So yeah, I think those changes and then save it immediately upon generation is probably the best bet.

But I do agree that some rework could be nice. Personally I'd like things consolidated a bit. Generation still uses one set off objects and tracker uses another set. Would be nice to sit of consolidate. For example, the old SMZ3 regions would keep the logic, but for rewards and medallions the generator could just use RegionInfo.

Unrelated, but I also wonder if the plando should use the item identifier used by the configs rather than the itemtype enum. That way people can refer to the template file.

**Comment by Vivelin** [comment]
> Unrelated, but I also wonder if the plando should use the item identifier used by the configs rather than the itemtype enum. That way people can refer to the template file.

Considering all the configuration is in the Tracking project, it'll be impossible to get to from the generation side of things without circular references and all that.

I definitely agree, though. I recall you talking about splitting off all the configuration into its own project somewhere; we could do this as part of that.

That said, I think this is in an okay state to release for testing.

**Comment by MattEqualsCoder** [comment]
Ah, yeah. That makes sense.

I'm looking at this now and doing testing, and a fresh plando yml file is still including the FileName property, which prevents it from loading successfully until it's removed from the yml file: 
image

**Comment by Vivelin** [comment]
This YAML library sucks.

**Review by MattEqualsCoder — APPROVED** [review]

File: src/Randomizer.SMZ3/Generation/PlandoConfig.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                .ToDictionary(x => x.ToString(), x => ((INeedsMedallion)x).Medallion);
+            Logic = world.Config.LogicConfig.Clone();
+        }
+
+        /// <summary>
+        /// Gets or sets the name of the file from which the plando config was
+        /// deserialized.
+        /// </summary>
+        [YamlIgnore]
+        public string FileName { get; set; }
```
[the remark this answers is not in the document]
Comment (Vivelin) [reply src/Randomizer.SMZ3/Generation/PlandoConfig.cs:47]: Yeah, just fixed it. I just swapped the YAML library out for another one which isn't broken and it no longer serializes the filename.

**Pull request merged**